### PR TITLE
Finish python runtime

### PR DESCRIPTION
**Pull request by endemics**
Probably buggy and not very elegant code, but this works for me(TM)

I've done most of the implementation with an eye on the javascript mock runtime so hopefully most of it is correct.

Note that during testing, I've noticed that there is a newer python runtime out there but made sure this implementation matched the runtime currently in the repository.

Should close half of #9

### CODE REVIEW

**Comment by vovanmix** [comment]
Please merge it, I need Python support in Lambda badly! 🙏

**Comment by cladden** [comment]
@mhart Do you have any issues with this PR?

**Comment by cladden** [comment]
If there are no issues, @neekolas could you help merge this PR?

**Comment by mhart** [comment]
Hey y'all – holiday time at the moment, but will hopefully find time to look over it during the next week

File: python2.7/run/runtime-mock.py [hunk 1]
```diff
@@ -1,41 +1,154 @@
...
+    print(*args, file=sys.stderr, **kwargs)
+
+def _random_account_id():
+  return random.randint(100000000000, 999999999999)
+
+def _random_invoke_id():
+    return str(uuid.uuid4())
+
+def _get_from_env(key, default):
+    if key in os.environ:
```
Comment (argos83) [inline python2.7/run/runtime-mock.py]: this function is equivalent of `os.environ.get(key, default)`

Comment (endemics) [inline python2.7/run/runtime-mock.py]: Thanks @argos83, I've updated the PR with the change to `os.environ.get`

Comment (billyshambrook) [inline python2.7/run/runtime-mock.py]: You could also do ```os.getenv(key, default)```

**Comment by cladden** [comment]
@mhart Are there any outstanding issues in this PR?

File: python2.7/run/runtime-mock.py [hunk 2]
```diff
@@ -1,41 +1,148 @@
...
 
 orig_stdout = sys.stdout
 orig_stderr = sys.stderr
 
-# TODO: finish this
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+def _random_account_id():
+  return random.randint(100000000000, 999999999999)
```
Comment (mhart) [inline python2.7/run/runtime-mock.py]: Will need to consistent with spacing – I personally prefer 2 spaces, but was sticking to PEP-8 which uses 4

File: python2.7/run/runtime-mock.py [hunk 3]
```diff
@@ -1,41 +1,148 @@
...
+
+def _random_invoke_id():
+    return str(uuid.uuid4())
+
+def _arn(region, account_id, fct_name):
+  return 'arn:aws:lambda:%s:%s:function:%s' % (region, account_id, fct_name)
```
Comment (mhart) [inline python2.7/run/runtime-mock.py]: Spacing here too

File: python2.7/run/Dockerfile [hunk 4]
```diff
@@ -20,7 +20,7 @@ RUN rm /var/runtime/awslambda/runtime.so
 ADD runtime-mock.py /var/runtime/awslambda/runtime.py
 
 # Not sure why permissions don't work just by modifying the owner
-RUN rm -rf /tmp && mkdir /tmp && chown -R sbx_user1051:495 /tmp && chmod 700 /tmp
+RUN rm -rf /tmp && mkdir /tmp && chown -R sbx_user1051:495 /tmp && chmod 700 /tmp && chmod -R 0777 /var/runtime
```
Comment (mhart) [inline python2.7/run/Dockerfile]: `/var/runtime` isn't writeable on Lambda – was there a reason for this change?

Comment (endemics) [inline python2.7/run/Dockerfile]: In the real lambda environment `/var/runtime/awslambda/runtime` is a compiled binary. In our case it's a .py file. We need the user we are running under to be able to create the `/var/runtime/awslambda/runtime.pyc` file.

Now I realise it's really hackish to give such wide permissions, and would be glad to implement a better solution. Ideas welcome :D

Comment (mhart) [inline python2.7/run/Dockerfile]: But doesn't it work just to not give those permissions? It just means the `.pyc` file isn't generated – but that's no big deal is it?

Comment (endemics) [inline python2.7/run/Dockerfile]: Mmh probably not, you're right. Especially since I guess the normal use case will be to run it as a one-off anyway.

**Comment by mhart** [comment]
Awesome, coming along really nicely! How about an example in the `examples` directory? Otherwise I guess we don't really have any way of testing it

**Comment by endemics** [comment]
> How about an example in the examples directory? Otherwise I guess we don't really have any way of testing it

Sure I'll add an example.

**Comment by joekiller** [comment]
I've tested this PR and the example and looks good for a merge to me.

**Comment by mhart** [comment]
@joekiller @endemics I just tested the example and I get no output at all.

After rebasing this branch off the latest master, in `python2.7/run`, I built the image:

```
docker build --pull -t python-test .
```

I then `cd`'ed to `examples/docker-run-python2.7` and ran:

```
docker run -v "$PWD":/var/task python-test
```

And there's no output.

**Comment by mhart** [comment]
It might be due to updates in `bootstrap.py` over the last few months?

Here's the diff for `/var/runtime/awslambda/bootstrap.py` (in `python2.7.tgz`) between the version in your branch and the latest master:

```diff
83c83
<     the module if it is already loaded, and it likely doesn't use any caches when
---
>     the module if it is already loaded, and it likely doesn't use any caches when
87c87
<
---
>
101c101,105
<                 request_handler = make_fault_handler(wsgi.FaultException("Cannot use built-in module {} as a handler module".format(modname), None, None))
---
>                 request_handler = make_fault_handler(wsgi.FaultException(
>                     "Cannot use built-in module {} as a handler module".format(modname),
>                     None,
>                     None
>                 ))
133,135c137,151
<
< def wait_for_start(ctrl_sock):
<     (invokeid, mode, handler, suppress_init, credentials) = lambda_runtime.recv_start(ctrl_sock)
---
> '''
> PYTHONPATH may have paths that were not available when the interpreter was launched.
> This would force the path importer cache get updated.
> '''
> def force_path_importer_cache_update():
>     for path in os.environ.get("PYTHONPATH", "").split(":"):
>         if path == os.environ["LAMBDA_RUNTIME_DIR"]:
>             continue
>         importer = sys.path_importer_cache.get(path, None)
>         if not importer or isinstance(importer, imp.NullImporter):
>             sys.path_importer_cache.pop(path, None)
>
> def wait_for_start():
>     (invokeid, mode, handler, suppress_init, credentials) = lambda_runtime.receive_start()
>     force_path_importer_cache_update()
138c154
<
---
>
141,142c157,169
< def wait_for_invoke(ctrl_sock):
<     (invokeid, data_sock, credentials, event_body, context_objs, invoked_function_arn) = lambda_runtime.receive_invoke(ctrl_sock)
---
> def wait_for_invoke():
>     (
>         invokeid,
>         data_sock,
>         credentials,
>         event_body,
>         context_objs,
>         invoked_function_arn,
>         trace_id,
>         parent_id,
>         sampled
>     ) = lambda_runtime.receive_invoke()
>
144a172,180
>     if trace_id:
>         xray_context = {
>             "trace_id" : trace_id,
>             "parent_id" : parent_id,
>             "sampled" : sampled
>         }
>
>         context_objs["xray_context"] = xray_context
>
157c193
<
---
>
202c238,239
<         context = LambdaContext(invokeid, context_objs, client_context, invoked_function_arn)
---
>         xray_context = context_objs.get('xray_context')
>         context = LambdaContext(invokeid, context_objs, client_context, xray_context, invoked_function_arn)
239c276,285
<         lambda_runtime.report_fault(invokeid, msgs[0], msgs[1], "Traceback (most recent call last):\n" + ''.join(traceback.format_list(tb_tuples)) + ''.join(traceback.format_exception_only(etype, value)))
---
>         lambda_runtime.report_fault(
>             invokeid,
>             msgs[0],
>             msgs[1],
>             (
>                 "Traceback (most recent call last):\n"
>                 + ''.join(traceback.format_list(tb_tuples))
>                 + ''.join(traceback.format_exception_only(etype, value))
>             )
>         )
241c287
<
---
>
280a327,329
> class XrayContext(object):
>     __slots__ = ['trace_id', 'parent_id', 'sampled']
>
282c331
<     def __init__(self, invokeid, context_objs, client_context, invoked_function_arn=None):
---
>     def __init__(self, invokeid, context_objs, client_context, xray_context, invoked_function_arn=None):
296c345,347
<
---
>         self.xray_context = make_obj_from_dict(XrayContext, xray_context)
>
>
319c370
<
---
>
334c385,388
<     logger_handler.setFormatter(logging.Formatter('[%(levelname)s]\t%(asctime)s.%(msecs)dZ\t%(aws_request_id)s\t%(message)s\n', '%Y-%m-%dT%H:%M:%S'))
---
>     logger_handler.setFormatter(logging.Formatter(
>         '[%(levelname)s]\t%(asctime)s.%(msecs)dZ\t%(aws_request_id)s\t%(message)s\n',
>         '%Y-%m-%dT%H:%M:%S'
>     ))
340,344c394,404
<     ctrl_sock = os.getenv('LAMBDA_CONTROL_SOCKET')
<     if ctrl_sock is None:
<         raise Exception("LAMBDA_CONTROL_SOCKET not set");
<
<     (invokeid, mode, handler, suppress_init, credentials) = wait_for_start(int(ctrl_sock))
---
>     # Remove lambda internal environment variables
>     for env in [
>         "_LAMBDA_CONTROL_SOCKET",
>         "_LAMBDA_SHARED_MEM_FD",
>         "_LAMBDA_LOG_FD",
>         "_LAMBDA_CONSOLE_SOCKET",
>         "_LAMBDA_RUNTIME_LOAD_TIME"
>     ]:
>         del os.environ[env]
>
>     (invokeid, mode, handler, suppress_init, credentials) = wait_for_start()
356c416
<         (invokeid, sockfd, credentials, event_body, context_objs, invoked_function_arn) = wait_for_invoke(int(ctrl_sock))
---
>         (invokeid, sockfd, credentials, event_body, context_objs, invoked_function_arn) = wait_for_invoke()
```

**Comment by mhart** [comment]
Two things jump out at me: `recv_start` has been renamed to `receive_start` and no longer takes an arg, and `receive_invoke` no longer takes an arg and returns `trace_id`, `parent_id`, `sampled` for the new `xray_context`

**Comment by mhart** [comment]
Alrighty – have fixed up all of the issues in https://github.com/lambci/docker-lambda/commit/5bb66cd83193dbcc8358f7814cd5fd0d7590bc03 – and released as 0.11.0

(I also changed all output, except for the result, to go to stderr – in line with the documentation on the README)

Thanks so much @endemics for all your work on this and being so patient!

**Pull request closed without merging**